c-scape/src/mem/string.rs

```diff
@@ -0,0 +1,318 @@
+use core::cell::SyncUnsafeCell;
+use core::ptr;
+use libc::{c_char, c_int};
+
+use crate::{malloc, memcpy};
+
+const NUL: c_char = 0;
+
+#[no_mangle]
+unsafe extern "C" fn stpcpy(mut d: *mut c_char, mut s: *const c_char) -> *mut c_char {
+    // Not defined in libc
+    // libc!(libc::stpcpy(d, s));
+
+    loop {
+        *d = *s;
+
+        if *d == NUL {
+            break;
+        }
+
+        d = d.add(1);
+        s = s.add(1);
+    }
+
+    d
+}
+
+#[no_mangle]
+unsafe extern "C" fn stpncpy(
+    mut d: *mut c_char,
+    mut s: *const c_char,
+    mut n: usize,
+) -> *mut c_char {
+    // Not defined in libc
+    // libc!(libc::stpncpy(d, s, n));
+
+    while n > 0 {
+        n = n - 1;
+
+        *d = *s;
+
+        if *d == NUL {
+            break;
+        }
+
+        d = d.add(1);
+        s = s.add(1);
+    }
+
+    for _ in 0..n {
+        *d = 0;
+        d = d.add(1);
+    }
+
+    d
+}
+
+#[no_mangle]
+unsafe extern "C" fn strcat(d: *mut c_char, s: *const c_char) -> *mut c_char {
+    libc!(libc::strcat(d, s));
+
+    strcpy(strchr(d, 0), s);
+    d
+}
+
+#[no_mangle]
+unsafe extern "C" fn strchr(s: *const c_char, c: c_int) -> *mut c_char {
+    libc!(libc::strchr(s, c));
+
+    let mut s = s as *mut c_char;
+    while *s != NUL {
+        if *s == c as _ {
+            return s;
+        }
+        s = s.add(1);
+    }
+
+    ptr::null_mut()
+}
+
+#[no_mangle]
+unsafe extern "C" fn strcmp(mut s1: *const c_char, mut s2: *const c_char) -> c_int {
+    libc!(libc::strcmp(s1, s2));
+
+    while *s1 != NUL && *s2 != NUL {
+        if *s1 != *s2 {
+            break;
+        }
+
+        s1 = s1.add(1);
+        s2 = s2.add(1);
+    }
+
+    (*s1 - *s2) as c_int
```

Comment:
This could misbehave on platforms where `c_char` is unsigned; it'd be good to cast to `c_schar` first. And I don't know Rust's integer promotion rules, but we may want to explicitly cast to `c_int` before doing the subtraction too.